### PR TITLE
12.0 mig sale commission product

### PR DESCRIPTION
**Pull request by scigghia**
first working sale_commission_product on V12

### CODE REVIEW

**Comment by pedrobaeza** [comment]
Please squash together all `OCA Transbot updated translations from Transifex` commits.

**Comment by scigghia** [comment]
> Please squash together all `OCA Transbot updated translations from Transifex` commits.

squashed

File: sale_commission_product/models/product.py
Comment (sergiocorato) [inline sale_commission_product/models/product.py]:
```suggestion
    _description = 'Product Product Agent'
```

File: sale_commission_product/models/product.py
Comment (sergiocorato) [inline sale_commission_product/models/product.py]:
```suggestion
    _description = 'Product Category Agent'
```

**Comment by sergiocorato** [comment]
On travis log I see a trailing whitespace and a weird reference of branch origin.

**Comment by sergiocorato** [comment]
@scigghia As said from @pedrobaeza, pls squash all commits of translation

**Comment by sergiocorato** [comment]
Travis red: https://travis-ci.org/github/OCA/commission/jobs/709510465#L296 and https://travis-ci.org/github/OCA/commission/jobs/709510465#L297 and https://travis-ci.org/github/OCA/commission/jobs/709510465#L300 origin

**Comment by scigghia** [comment]
> Travis red: https://travis-ci.org/github/OCA/commission/jobs/709510465#L296 and https://travis-ci.org/github/OCA/commission/jobs/709510465#L297 and https://travis-ci.org/github/OCA/commission/jobs/709510465#L300 origin

done, but i'm in doubt for https://travis-ci.org/github/OCA/commission/jobs/709510465#L300

**Comment by sergiocorato** [comment]
> > Travis red: https://travis-ci.org/github/OCA/commission/jobs/709510465#L296 and https://travis-ci.org/github/OCA/commission/jobs/709510465#L297 and https://travis-ci.org/github/OCA/commission/jobs/709510465#L300 origin
> 
> done, but i'm in doubt for https://travis-ci.org/github/OCA/commission/jobs/709510465#L300

Me too, as :green_apple: no worry

**Comment by FrancoTampieri** [comment]
I review the code and I don't find any problem, for me is good +1

**Review by FrancoTampieri — APPROVED** [review]
Is ok for me, release is Good +1

**Comment by sergiocorato** [comment]
Tests :pray:?

File: sale_commission_product/models/product.py
Comment (sergiocorato) [inline sale_commission_product/models/product.py]: Isn't it a @api.model?

File: sale_commission_product/models/product.py
Comment (sergiocorato) [inline sale_commission_product/models/product.py]: As agent is always passed, why searching without agent if below we search by agent? Better search by agent and only if not found search a second time without it.

File: sale_commission_product/models/product.py
Comment (sergiocorato) [inline sale_commission_product/models/product.py]: Something like this could be better?
```
commission_ids = self.search_read([('product_id', '=', product), ('agent', '=', agent.id)], ['commission'])
if not commission_ids:
    commission_ids = self.search_read([('product_id', '=', product)], ['commission'])
if commission_ids:
   commission_id = commission_ids[0].id
```

File: sale_commission_product/models/product.py
Comment (sergiocorato) [inline sale_commission_product/models/product.py]: If `product` is an id, rename it `product_id` to be more clear
Why passing agent as a recordset?

File: sale_commission_product/models/product.py
Comment (sergiocorato) [inline sale_commission_product/models/product.py]: Same remarks of `get_commission_id_product` plus call this from it

File: sale_commission_product/models/product.py
Comment (sergiocorato) [inline sale_commission_product/models/product.py]: agent_ids + oldname=agents + migration script

File: sale_commission_product/models/product.py
Comment (sergiocorato) [inline sale_commission_product/models/product.py]: agent_id + oldname=agent + migration script

File: sale_commission_product/models/product.py
Comment (sergiocorato) [inline sale_commission_product/models/product.py]: commission_id + oldname=commission + migration script

File: sale_commission_product/models/product.py
Comment (sergiocorato) [inline sale_commission_product/models/product.py]: agent_id + oldname=agent + migration script

File: sale_commission_product/models/product.py
Comment (sergiocorato) [inline sale_commission_product/models/product.py]: commission_ids + oldname=commission + migration script

File: sale_commission_product/models/sale_order.py
Comment (sergiocorato) [inline sale_commission_product/models/sale_order.py]: Better put this logic in a unique function that return always `commission_id` like:
```
commission_id = self.env['product.product.agent'].get_commission_id(self.product_id, agent)
```

File: sale_commission_product/tests/test_commission_product.py
Comment (sergiocorato) [inline sale_commission_product/tests/test_commission_product.py]: removing `agent` would give the same result or not?

**Review by sergiocorato — CHANGES_REQUESTED** [review]

File: sale_commission_product/models/product.py
Comment (scigghia) [reply sale_commission_product/models/product.py]: done , but i suppose there is no need for migration script

File: sale_commission_product/tests/test_commission_product.py
Comment (scigghia) [reply sale_commission_product/tests/test_commission_product.py]: no, is a category with a specific agent that will override every default settings on customer o product

File: sale_commission_product/views/product_view.xml
Comment (brian10048) [inline sale_commission_product/views/product_view.xml:3]: The commission view only shows on product variant form, but not on the main product template form

**Go to Sales > Products > Products**
The commissions section does not appear in the Product form
Screenshot from 2020-11-10 11-25-29


**Go to Sales > Products > Product Variants**
The commissions section is in the view now.
Screenshot from 2020-11-10 11-22-20


**Recommend to duplicate the view and inherit `product.product_template_only_form_view`**

File: sale_commission_product/__manifest__.py
Comment (rousseldenis) [inline sale_commission_product/__manifest__.py]: This should go into CONTRIBUTORS.rst readme fragment.

If you want to set maintainers, please use 'maintainers' key with github login.

File: sale_commission_product/__manifest__.py
Comment (rousseldenis) [inline sale_commission_product/__manifest__.py]: As it is a migration, please reset to 12.0.1.0.0

File: sale_commission_product/__manifest__.py
Comment (rousseldenis) [inline sale_commission_product/__manifest__.py:6]:
```suggestion
    'name': 'Sale commission product',
```

File: sale_commission_product/models/account_invoice.py
Comment (rousseldenis) [inline sale_commission_product/models/account_invoice.py]: Please put self.env... before the loop as variable.

File: sale_commission_product/models/account_invoice.py
Comment (rousseldenis) [inline sale_commission_product/models/account_invoice.py]: Same as above

File: sale_commission_product/models/product.py
Comment (rousseldenis) [inline sale_commission_product/models/product.py:18]: Please put one model per file.

File: sale_commission_product/models/product.py
Comment (rousseldenis) [inline sale_commission_product/models/product.py:36]: Why search_read() ?

File: sale_commission_product/models/product.py
Comment (rousseldenis) [inline sale_commission_product/models/product.py]: Please put fields on top

File: sale_commission_product/models/product.py
Comment (rousseldenis) [inline sale_commission_product/models/product.py]: Why this ?

File: sale_commission_product/models/product.py
Comment (rousseldenis) [inline sale_commission_product/models/product.py:88]: IMHO, the code to retrieve commission from product or category is quite duplicated. I think this should be reworked.

File: sale_commission_product/models/product.py
Comment (rousseldenis) [inline sale_commission_product/models/product.py]: Please put fields on top

**Review by rousseldenis — CHANGES_REQUESTED** [review]
Code review

**Comment by josepmy** [comment]
Hi,
Just I downloaded it to test together with oca sale_commision module, so I need to assign different commission between categories.
After install the module if I try create a new order and add any product I get this error:

> La operación no se puede completar:
> Crear / actualizar: no se establece un campo obligatorio.
> Eliminar: otro modelo requiere que se elimine el registro. Si es posible, archívelo en su lugar.
> 
> Modelo sale.order.line.agent (sale.order.line.agent), Campo: Agent (agent)`


From console

`2020-11-25 15:48:15,935 17634 ERROR margi-pro odoo.sql_db: bad query: INSERT INTO "sale_order_line_agent" ("id", "create_uid", "create_date", "write_uid", "write_date", "object_id") VALUES (nextval('sale_order_line_agent_id_seq'), 2, (now() at time zone 'UTC'), 2, (now() at time zone 'UTC'), 30095) RETURNING id
ERROR: null value in column "agent" violates not-null constraint
DETAIL:  Failing row contains (31, 30095, null, null, null, 2, 2020-11-25 15:48:15.895789, 2, 2020-11-25 15:48:15.895789)`

Uninstalling the module the sale_commision work fine again.

**Comment by josepmy** [comment]
@scigghia I debug the code and I think that in `def product_id_change(self):` don't receive the agents ID but I'm blocked. Can you test it?

**Comment by josepmy** [comment]
@scigghia with this code I can save but the commission isn't working.
`class SaleOrderLine(models.Model):
    _inherit = "sale.order.line"

    @api.model
    def create(self, vals):
        """Add agents for records created from automations instead of UI."""
        # We use this form as this is the way it's returned when no real vals
        agents_vals = vals.get('agents', [(6, 0, [])])
        # if agents_vals and agents_vals[0][0] == 6 and not agents_vals[0][2]:
        if agents_vals and not agents_vals[0][2]:
            vals['agents'] = self._prepare_agents_vals(vals=vals)
        return super().create(vals)`

File: sale_commission_product/views/product_view.xml
Comment (scigghia) [reply sale_commission_product/views/product_view.xml:3]: Hi @brian10048 I think we could consider only product.product model because, this is the model linked in the sale order lines.
And the template cloud have several variants (product.product) so i think che commission i better to stay on product.product.

have you other use case to consider? would you add a pull request above mine ?
thank you

File: sale_commission_product/__manifest__.py
Comment (scigghia) [reply sale_commission_product/__manifest__.py]: got it

File: sale_commission_product/models/product.py
Comment (scigghia) [reply sale_commission_product/models/product.py:36]: it's faster

File: sale_commission_product/models/product.py
Comment (scigghia) [reply sale_commission_product/models/product.py]: what? empty string?

**Comment by scigghia** [comment]
hi @josepmy i've tested this pr with an empty database.
i created two categories, two product, one agent with different commission on each category and in works

could you give me more details of your issue?

**Comment by josepmy** [comment]
Hi @scigghia <https://github.com/scigghia> my issue is when you have two agents on the same sale order.

My case is:

Create two agents.
Agent A have commission over all products
Agent B have 0% commision over one category.

Mix products in one sale and try to save.
Open existing sale order and add one product of the category of agent B.

**Comment by scigghia** [comment]
Has the Customer those two agents?

**Comment by josepmy** [comment]
yes

**Comment by scigghia** [comment]
I created a new PR here https://github.com/OCA/commission/pull/288
i'm trying to rework this module
could you take a look?

**Comment by josepmy** [comment]
Of course

**Comment by pedrobaeza** [comment]
Can this one be closed then?

**Comment by scigghia** [comment]
yes @pedrobaeza 
@josepmy i'm waiting for your feedback here https://github.com/OCA/commission/pull/288

**Comment by josepmy** [comment]
ok

**Pull request closed without merging**
